lib/ecto/multi.ex

```diff
@@ -0,0 +1,189 @@
+defmodule Ecto.Multi do
+  alias __MODULE__
+  alias Ecto.Changeset
+
+  defstruct operations: [], names: MapSet.new
+
+  @type run :: (t -> {:ok | :error, any}) | {module, atom, [any]}
+  @typep operation :: {:changeset, Changeset.t, Keyword.t} |
+                      {:run, run} |
+                      {:update_all, Ecto.Query.t, Keyword.t} |
+                      {:delete_all, Ecto.Query.t, Keyword.t}
+  @type name :: atom
+  @opaque t :: %__MODULE__{operations: [{name, operation}], names: MapSet.t}
+
+  @spec new :: t
+  def new do
+    %Multi{}
+  end
+
+  @spec append(t, t) :: t
+  def append(lhs, rhs) do
+    merge(lhs, rhs, &(&2 ++ &1))
+  end
+
+  @spec prepend(t, t) :: t
+  def prepend(lhs, rhs) do
+    merge(lhs, rhs, &(&1 ++ &2))
+  end
+
+  def merge(%Multi{names: names1, operations: ops1} = lhs,
+            %Multi{names: names2, operations: ops2} = rhs,
+            joiner) do
+    if MapSet.disjoint?(names1, names2) do
+      %Multi{names: MapSet.union(names1, names2),
+             operations: joiner.(ops1, ops2)}
+    else
+      common = MapSet.intersection(names1, names2) |> MapSet.to_list
+      raise """
+      when merging following Ecto.Multi:
+
+      #{inspect lhs}
+
+      #{inspect rhs}
+
+      both declared operations: #{inspect common}
+      """
+    end
+  end
+
+  @spec insert(t, name, Changeset.t | Ecto.Schema.t) :: t
+  def insert(multi, name, changeset_or_struct, opts \\ [])
+
+  def insert(multi, name, %Changeset{} = changeset, opts) do
+    add_changeset(multi, :insert, name, changeset, opts)
+  end
+
+  def insert(multi, name, struct, opts) do
+    insert(multi, name, Changeset.change(struct), opts)
+  end
+
+  @spec update(t, name, Changeset.t) :: t
+  def update(multi, name, %Changeset{} = changeset, opts \\ []) do
+    add_changeset(multi, :update, name, changeset, opts)
+  end
+
+  @spec delete(t, name, Changeset.t | Ecto.Schema.t) :: t
+  def delete(multi, name, changeset_or_struct, opts \\ [])
+
+  def delete(multi, name, %Changeset{} = changeset, opts) do
+    add_changeset(multi, :delete, name, changeset, opts)
+  end
+
+  def delete(multi, name, struct, opts) do
+    delete(multi, name, Changeset.change(struct), opts)
+  end
+
+  defp add_changeset(multi, action, name, changeset, opts) when is_list(opts) do
+    add_operation(multi, name, {:changeset, put_action(changeset, action), opts})
+  end
+
+  defp put_action(%{action: nil} = changeset, action) do
+    %{changeset | action: action}
+  end
+
+  defp put_action(%{action: action} = changeset, action) do
+    changeset
+  end
+
+  defp put_action(%{action: original}, action) do
+    raise ArgumentError, "you provided a changeset with an action already set " <>
+      "to #{inspect original} when trying to #{action} it"
+  end
+
+  @spec run(t, name, run) :: t
+  def run(multi, name, run) when is_function(run, 1) do
+    add_operation(multi, name, {:run, run})
```

Comment:
It's problematic because I can't easily add an argument to the run function in that case:

```elixir
  defp apply_operation({:run, {mod, fun, args}}, acc, _repo),
    do: apply(mod, fun, [acc | args])
```

Or am I missing something? Is there a way to call a fun as mfa and add additional argument later?